Enforce @safe or @system attributes on unittests

wilzbach opened the pull request:
From #4947 we learned that it might be a good idea to remind the submitter and reviewer that all `unittest` blocks should be annotated.

edit: this PR uses a new [Dscanner plugin](https://github.com/Hackerpilot/Dscanner/pull/389) to enforce `@system` or `@safe` on unittest blocks.

~~~The idea of this PR is to automate this in a simple, but yet effective way by checking for the pattern `^unittest`.~~~

~~~There were only 32 `unittest` blocks in the entire Phobos codebase that did use two lines for the attributes ([see the separate commit](https://github.com/dlang/phobos/commit/29ecc5d32c4355c4f84dac1882aeafb7562c3713). While it's definitely possible to write a smarter checker, for the greater good of having a simple way to check/grep for attributes I unified the style.~~~

~~~For all other `unittest` blocks without attributes I applied `sed`. As you can see that after the `@system/@safe` [annotation games](https://github.com/dlang/phobos/pulls?utf8=%E2%9C%93&q=safe%20system) after the [Call to Action: making Phobos @safe](http://forum.dlang.org/post/nkn1g6$b00$1@digitalmars.com) this summer and more a lot of new changes without annotation managed to sneak in, hence the idea of taking a snapshot of the current state (aka `@system`) and thus be able to enforce it for new changes.~~~

~~~One might also argue that it's easier to find unittest blocks that haven't been annotated with `@safe/@system` yet, but  (1) since the annotation wars in summer the interest in continuing this on a steady basis quickly dropped, (2) it's still possible to quickly grep for `@system` and continue to improve D's safety (in the past Phobos was "safetyed" on a per-module basis anyways), (3) I was astonished by how few `unittest`s blocks were left (see the diff yourself), (4) it's a small noise "for the greater good" on our journey to reduce potential problems with new changes. All new changes can now be enforced to be properly annotated :)~~~

This PR is only possible due to the hard work of @atilaneves, @BBasile and all the other awesome Phobos contributors!

ZombineDev commented on posix.mak:
```diff
@@ -511,6 +511,9 @@ style: ../dscanner/dsc
 	grep -nr 'do *{$$' $$(find . -name '*.d') ; test $$? -eq 1
 
+	@echo "Enforce attributes for unittests. Test should at least be annotated with either @safe or @system"
+	grep -nr '^unittest' $$(find . -name '*.d') ; test $$? -eq 1
```
It also seems like that it doesn't look at unittests of member functions.

wilzbach commented:
> A simple algorithm to do things more robustly goes like this:

Yeah I was lazy. However, it's not that to add a check to Dscanner. So there's no need to reinvent the wheel :)

https://github.com/Hackerpilot/Dscanner/pull/389

ZombineDev commented:
> Hackerpilot/Dscanner#389

Even better!

wilzbach commented:
> Hackerpilot/Dscanner#389

So with the style checking being part of Dscanner, I updated this PR to
- enable the new Dscanner check
- use `@system` for all non-annotated unittests

For the latter part I wrote a simple script that parses the Dscanner output and fixes the warning by stupidly applying `@system` (see below in case someone is interested).
@CyberShadow: I am pretty sure that there's a way to use digger to rewrite `@system unittest` blocks to `@safe` while compilation is successful, but imho this goes beyond the scope of this PR.

```d
#!/usr/bin/env rdmd

void main(string[] args)
{
    import std.algorithm, std.array, std.ascii, std.conv, std.file, std.range, std.regex, std.stdio, std.string;
    string inFileList = args[1];

    struct Entry
    {
        size_t line;
        size_t column;
    }

    Entry[][string] replacements;
    auto re = regex(`(.+)\(([0-9]+):([0-9]+)\)`);
    // sortedness from dscanner is expected
    foreach (entry; File(inFileList).byLineCopy)
    {
        auto m = entry.matchFirst(re);
        if (m.length < 3)
            continue;
        const key = m[1];
        replacements[key] ~= Entry(m[2].to!size_t, m[3].to!size_t);
    }

    replacements.writeln;

    foreach (key, values; replacements.byPair)
    {
        auto tmpFile = tempDir ~ "/" ~ key.replace("/", "_");
        auto outFile = File(tmpFile, "w");
        writeln("open: ", key);
        auto lines = File(key).byLineCopy.array;
        foreach (entry; values)
        {
            auto line = entry.line - 1; // dscanner lines start with 1
            if (!lines[line].canFind("unittest"))
            {
                writeln("dscanner error in %s", line);
                continue;
            }
            auto parts = lines[line].split("unittest");
            lines[line] = parts[0] ~ "@system unittest" ~ parts[1];
        }

        // dump file
        foreach (line; lines)
            outFile.writeln(line);

        outFile.flush;
        tmpFile.copy(key);
        tmpFile.remove;
    }
}
```

In case anyone is interested - it can be run with:

```
> make -f posix.mak style | grep warn > result
rdmd -g ./replacement.d result
```

CyberShadow commented:
> @CyberShadow: I am pretty sure that there's a way to use digger to rewrite @system unittest blocks to @safe while compilation is successful, but imho this goes beyond the scope of this PR.

Sorry, I don't understand. Did you mean DustMite?

wilzbach commented:
> Sorry, I don't understand. Did you mean DustMite?

Oh yep - sorry. I was just trying to say that it's possible to build automation to test every `unittest` block for `@safe` or `@system`, but this automation is independent from this PR.

Btw rebased and added `@safe` and `@system` to new unittests blocks from recent commits.

wilzbach commented:
(superseded by #5166 which checks the for all `unittests` automatically whether they can be `@safe` without changes)
